src/t_hash.c

```diff
@@ -1252,13 +1264,44 @@ void hsetexCommand(client *c) {
 
     bool has_volatile_fields = hashTypeHasVolatileFields(o);
 
-    /* In case we are expiring all the elements prepare a new argv since we are going to delete all the expired fields. */
+    /* Prepare a new argv when rewriting the command. If set_expired is true,
+     * all expired fields will be deleted. Otherwise, if rewriting is needed due to NX/XX/FNX/FXX flags,
+     * copy the command, key, and optional arguments, skipping the NX/XX/FNX/FXX flags. */
     if (set_expired) {
         new_argv = zmalloc(sizeof(robj *) * (num_fields + 2));
         new_argv[new_argc++] = shared.hdel;
         incrRefCount(shared.hdel);
         new_argv[new_argc++] = c->argv[1];
         incrRefCount(c->argv[1]);
+    } else if (need_rewrite_for_nx_xx_fnx_fxx_ex_px_exat) {
+        /* We use new_argv for rewrite */
+        new_argv = zmalloc(sizeof(robj *) * c->argc);
+        int j = 0;
+        // Command
+        new_argv[j++] = c->argv[0];
+        incrRefCount(c->argv[0]);
+        // Key
+        new_argv[j++] = c->argv[1];
+        incrRefCount(c->argv[1]);
+        // Copy optional args (skip NX/XX/FNX/FXX)
+        for (int i = 2; i < fields_index; i++) {
+            if (strcmp(c->argv[i]->ptr, "NX") &&
+                strcmp(c->argv[i]->ptr, "XX") &&
+                strcmp(c->argv[i]->ptr, "FNX") &&
+                strcmp(c->argv[i]->ptr, "FXX")) {
+                /* Propagate as HSETEX Key Value PXAT millisecond-timestamp if there is
+                 * EX/PX/EXAT flag. */
+                if (expire && !(flags & ARGS_PXAT) && c->argv[i + 1] == expire) {
+                    robj *milliseconds_obj = createStringObjectFromLongLong(when);
+                    rewriteClientCommandArgument(c, i, shared.pxat);
+                    rewriteClientCommandArgument(c, i + 1, milliseconds_obj);
+                    decrRefCount(milliseconds_obj);
+                }
+                new_argv[j++] = c->argv[i];
+                incrRefCount(c->argv[i]);
+            }
+        }
+        new_argc = j;
```

Comment:
Maybe little bit cleaner this way?
```suggestion
        // Copy optional args (skip NX/XX/FNX/FXX)
        for (int i = 0; i < fields_index; i++) {
            if (strcmp(c->argv[i]->ptr, "NX") &&
                strcmp(c->argv[i]->ptr, "XX") &&
                strcmp(c->argv[i]->ptr, "FNX") &&
                strcmp(c->argv[i]->ptr, "FXX")) {
                /* Propagate as HSETEX Key Value PXAT millisecond-timestamp if there is
                 * EX/PX/EXAT flag. */
                if (expire && !(flags & ARGS_PXAT) && c->argv[i + 1] == expire) {
                    robj *milliseconds_obj = createStringObjectFromLongLong(when);
                    new_argv[new_argc++] = shared.pxat;
                    new_argv[new_argc++] = milliseconds_obj;
                } else {
                    new_argv[new_argc++] = c->argv[i];
                    incrRefCount(c->argv[i]);
                }
            }
        }
```